redash/models/__init__.py

```diff
@@ -138,28 +191,40 @@ def get_by_id(cls, _id):
         return cls.query.filter(cls.id == _id).one()
 
     def delete(self):
+        # Delete the relevant metadata about a data source first.
+        tables = TableMetadata.query.filter(TableMetadata.data_source_id == self.id).options(load_only('id'))
+        ColumnMetadata.query.filter(
+            ColumnMetadata.table_id.in_(tables.subquery())
+        ).delete(synchronize_session=False)
+
+        tables.delete()
+
         Query.query.filter(Query.data_source == self).update(dict(data_source_id=None, latest_query_data_id=None))
         QueryResult.query.filter(QueryResult.data_source == self).delete()
         res = db.session.delete(self)
         db.session.commit()
         return res
 
     def get_schema(self, refresh=False):
-        key = "data_source:schema:{}".format(self.id)
-
-        cache = None
-        if not refresh:
-            cache = redis_connection.get(key)
-
-        if cache is None:
-            query_runner = self.query_runner
-            schema = sorted(query_runner.get_schema(get_stats=refresh), key=lambda t: t['name'])
-
-            redis_connection.set(key, json_dumps(schema))
-        else:
-            schema = json_loads(cache)
-
-        return schema
+        schema = []
+        tables = TableMetadata.query.filter(TableMetadata.data_source_id == self.id).all()
+        for table in tables:
+            table_info = {
+                'name': table.table_name,
+                'exists': table.table_exists,
```

Comment:
We should probably show only existing tables/columns in the UI. We persist them in the DB, so no data is lost, but I don't see much value in showing them to the user.

Reply:
They were being filtered on the front-end initially but now they get filtered out in the backend